components/ws-daemon/pkg/cgroup/plugin_process_priority_v2.go

```diff
@@ -0,0 +1,130 @@
+	"syscall"
+	"time"
+
+	"github.com/shirou/gopsutil/process"
+	"golang.org/x/xerrors"
+
+	"github.com/gitpod-io/gitpod/common-go/log"
+)
+
+// ProcessType referes to the kinds of prioritisable processes in the cgroup
+type ProcessType string
+
+const (
+	// ProcessSupervisor referes to a supervisor process
+	ProcessSupervisor ProcessType = "supervisor"
+	// ProcessIDE refers to node.js IDE process
+	ProcessIDE ProcessType = "ide"
+	// ProcessDefault referes to any process that is not one of the above
+	ProcessDefault ProcessType = "default"
+)
+
+type ProcessPriorityV2 struct {
+	ProcessPriorities map[ProcessType]int
+}
+
+func (c *ProcessPriorityV2) Name() string  { return "process-priority-v2" }
+func (c *ProcessPriorityV2) Type() Version { return Version2 }
+
+func (c *ProcessPriorityV2) Apply(ctx context.Context, basePath, cgroupPath string) error {
+	fullCgroupPath := filepath.Join(basePath, cgroupPath)
+
+	_, err := os.Stat(fullCgroupPath)
+	if os.IsNotExist(err) {
+		return xerrors.Errorf("cannot read cgroup directory %s: %w", fullCgroupPath, err)
+	}
+
+	go func() {
+		time.Sleep(10 * time.Second)
+
+		data, err := ioutil.ReadFile(filepath.Join(fullCgroupPath, "workspace", "user", "cgroup.procs"))
+		if err != nil {
+			log.WithField("path", fullCgroupPath).WithError(err).Errorf("cannot read cgroup.procs file")
+			return
+		}
+
+		for _, line := range strings.Split(string(data), "\n") {
+			line = strings.TrimSpace(line)
+			if len(line) == 0 {
+				continue
+			}
+
+			pid, err := strconv.ParseInt(line, 10, 64)
+			if err != nil {
+				log.WithError(err).WithField("line", line).Warn("cannot parse pid")
+				continue
+			}
+
+			proc, err := process.NewProcess(int32(pid))
+			if err != nil {
+				if errors.Is(err, process.ErrorProcessNotRunning) {
+					continue
+				}
+
+				log.WithError(err).WithField("pid", pid).Warn("cannot get process")
+				continue
+			}
+
+			procType := determineProcessType(proc)
+			if procType == ProcessDefault {
+				continue
+			}
+
+			priority, ok := c.ProcessPriorities[procType]
+			if !ok {
+				continue
+			}
+
+			cmd, _ := proc.Cmdline()
+
+			log.WithField("pid", pid).WithField("priority", priority).WithField("process", cmd).Debug("changing process priority")
+
+			err = syscall.Setpriority(syscall.PRIO_PROCESS, int(pid), priority)
+			if err != nil {
+				log.WithError(err).WithField("pid", pid).WithField("priority", priority).Warn("cannot set process priority")
+			}
+		}
+	}()
+
+	return nil
+}
+
+type processTypeIndicator interface {
+	CmdlineSlice() ([]string, error)
+}
+
+func determineProcessType(p processTypeIndicator) ProcessType {
+	cmdLine, err := p.CmdlineSlice()
+	if err != nil || len(cmdLine) == 0 {
+		return ProcessDefault
+	}
+
+	cmd := cmdLine[0]
+	if (cmd == "/bin/bash" || cmd == "sh") && len(cmdLine) > 1 {
+		cmd = cmdLine[1]
+	}
+
+	if strings.HasSuffix(cmd, "supervisor") {
+		return ProcessSupervisor
+	}
+
+	if strings.HasSuffix(cmd, "gitpod-code") {
+		return ProcessIDE
+	}
```

Comment:
This seems too easy to fake. How about adding this below condition?
- `supervisor` should have no parent process
- `gitpod-code`'s parent must be `supervisor`

Reply:
Good point. Fixed